partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/CoverageCompiler.kt

```diff
@@ -0,0 +1,437 @@
+
+        val thunk = nestCompilationContext(ExpressionContext.NORMAL, emptySet()) {
+            compileAstStatement(transformedAst)
+        }
+
+        return object : Expression {
+            override val coverageStructure: CoverageStructure = CoverageStructure(
+                branches = branches.toMap(),
+                branchConditions = conditions.toMap()
+            )
+
+            override fun eval(session: EvaluationSession): ExprValue {
+                val env = Environment(
+                    session = session,
+                    locals = session.globals,
+                    current = session.globals,
+                    branchConditionCounts = mutableMapOf(),
+                    branchCounts = mutableMapOf()
+                )
+                return thunk(env)
+            }
+
+            override fun evaluate(session: EvaluationSession): PartiQLResult {
+                val env = Environment(
+                    session = session,
+                    locals = session.globals,
+                    current = session.globals,
+                    branchConditionCounts = mutableMapOf(),
+                    branchCounts = mutableMapOf()
+                )
+                val value = thunk(env)
+                return PartiQLResult.Value(
+                    value = value,
+                    coverageData = {
+                        CoverageData(
+                            branchConditionCount = env.branchConditionCounts!!.toMap(),
+                            branchCount = env.branchCounts!!.toMap()
+                        )
+                    },
+                    coverageStructure = { coverageStructure }
+                )
+            }
+        }
+    }
+/**
+     override fun compile(originalAst: PartiqlAst.Statement): Expression {
+     contextStack.push(Context.NOT_IN_BRANCH)
+     val expression = super.compile(originalAst)
+     return object : Expression {
+     override val coverageStructure: CoverageStructure = CoverageStructure(
+     branches = branches.toMap(),
+     branchConditions = conditions.toMap()
+     )
+
+     @Suppress("DEPRECATION")
+     override fun eval(session: EvaluationSession): ExprValue {
+     return expression.eval(session)
+     }
+
+     override fun evaluate(session: EvaluationSession): PartiQLResult {
+     val result = (expression.evaluate(session) as PartiQLResult.Value).value
+     return PartiQLResult.Value(
+     value = result,
+     coverageData = {
+     val returnedConditionCounts = conditionCounts.toMap()
+     val returnedBranchCounts = branchCounts.toMap()
+     // conditionCounts.replaceAll { _, _ -> 0 }
+     // branchCounts.replaceAll { _, _ -> 0 }
+     CoverageData(branchConditionCount = returnedConditionCounts, branchCount = returnedBranchCounts)
+     },
+     coverageStructure = { coverageStructure }
+     )
+     }
+     }
+     }
+     **/
+
+    // TODO: Figure out how we can determine whether an ID should be a boolean.
```

Comment:
Could you please elaborate this and the next comment? I guess I'm missing a prior context.